internal/appsec/waf/waf.go

```diff
@@ -944,3 +968,41 @@ func cindexCharPtrArray(array **C.char, i int) *C.char {
 	base := uintptr(unsafe.Pointer(array))
 	return *(**C.char)(unsafe.Pointer(base + unsafe.Sizeof((*C.char)(nil))*uintptr(i)))
 }
+
+// Atomic reference counter helper initialized at 1 so that 0 is the special
+// value meaning that the object was released and is no longer usable.
+type atomicRefCounter uint32
+
+func (c *atomicRefCounter) unwrap() *uint32 {
+	return (*uint32)(c)
+}
+
+// CAS implementation in order to avoid changing the counter when 0 has been
```

Comment:
Not sure why it says CAS implementation for all three method, isn't CAS Compare And Swap?

Reply:
CAS spin-loop indeed (let me further clarify it in the code). And I was emphasising that `increment` and `decrement` must be CAS to avoid bypassing this 0 limit.

Reply:
Clarified in a46450ef